doc/tutorial/concepts/index.md

```diff
@@ -0,0 +1,311 @@
+# C++ Concepts {#tutorial_concepts}
+
+[TOC]
+
+This tutorial introduces "C++ Concepts", a feature of C++20 (and available to some extent in older GCC versions).
+You will learn the terminology used in the context of concepts and how to use SeqAn3's concepts in your application.
+
+\tutorial_head{Moderate, 60 min, \ref setup\, \ref tutorial_argument_parser,}
+
+This tutorial teaches the very basics of working with concepts. For more background and information on how to implement
+your own concepts, we recommend:
+  * A well-readable [paper](http://www.stroustrup.com/good_concepts.pdf) with motivation and historical background.
+  * The (rather dense) [documentation on cppreference](https://en.cppreference.com/w/cpp/language/constraints).
+
+# Constraints
+
+## Motivation
+
+Generic programming is all about making your algorithms work with different types, e.g. the local alignment algorithm
```

Comment:
One central design goal of SeqAn is to provide generic algorithms and data structures which can be used for different types without reimplementing the same algorithms over and over again for particular types. This has multiple benefits, among which the maintainability will be improved due to an additional level of abstraction and most importantly the code can be reused with user provided types. A familiar example for generic code is the std::vector and the algorithms in the STL. The vector and the algorithms are templatised. This means that they do not depend on a specific type but you could rather put in any type. This sounds awesome, but very often this is not completely true. While the given type can be used arbitrarily it very often must offer a specific interface. To give an example in SeqAn consider the local alignment algorithm. This algorithm computes the best local match between two sequences over a finite alphabet. The algorithm is generic as such that it allows any alphabet type as long as this type offers a minimal interface which is used inside of the algorithm (e.g. objects of the alphabet type must be equality comparable). 
Before C++20, this couldn't be checked easily such that using the interface with non-conformant types would result in very hard to read error messages emitted by the compiler and further would lead to a huge frustration by the user. In the following part of the tutorial you will learn how to *constraint* such template arguments of generic functions and data structures and how this can have a huge benificial impact on your code quality and quantity.

Reply:
Thanks for this text, I have imported it with few changes.